### PR TITLE
add StartupAllowedCPUs= and StartupAllowedMemoryNodes= directives

### LINKED ISSUE #20733: RFE: enable control of cgroup attributes at runtime
**Is your feature request related to a problem? Please describe.**

We have a system where fast re(boots) are a requirement, thus I want to be able to give all available resources to the system.slice  during boot and shutdown. However after boot (after some arbitrary target is reached) I would like to ensure processes in the system.slice cgroup only run on a subset of available CPUs (to avoid interrupting another latency sensitive application in another slice). We can use the AllowedCPUs option to pin a cgroup to a cpuset, however this requires either of:

   * AllowedCPUs is set for the system slice during boot (this slows our boot down so isn't an option).
   * A runtime drop-in is created to change AllowedCPUs at runtime. This seems like a heavy weight solution and possibly difficult to manage by another user defined service.
   * We don't set AllowedCPUs for the system.slice, rather after boot we manually twiddle the cpuset.cpus file for the system.slice. This violates the principle of a "single writer" that systemd preaches, sadly however this is the solution I am currently using.

**Describe the solution you'd like**

A less heavy weight way (as opposed to runtime drop-in creation and a daemon-reload) of changing resource configuration of a service/slice/scope. These properties are currently dbus RO, if they could be made RW then this would be a lighter weight way to change configuration. I'd be happy to work on a patch for this.

**Describe alternatives you've considered**

Copy/paste from above:

   * AllowedCPUs is set for the system slice during boot (this slows our boot down so isn't an option).
   * A runtime drop-in is created to change AllowedCPUs at runtime. This seems like a heavy weight solution and possibly difficult to manage by another user defined service.
   * We don't set AllowedCPUs for the system.slice, rather after boot we manually twiddle the cpuset.cpus file for the system.slice. This violates the principle of a "single writer" that systemd preaches, sadly however this is the solution I am currently using.

**The systemd version you checked that didn't have the feature you are asking for**

v249.

### PR DESCRIPTION
**Pull request by pdmorrow**
By adding StartupAllowedCPUs and StartupAllowedMemoryNodes we can control CPU affinity during boot in a different manner than during normal runtime. This is useful when boot time matters as much as post boot CPU affinity configuration.

Fixes #20733

### CODE REVIEW

File: src/core/cgroup.c [hunk 1]
```diff
@@ -387,6 +389,7 @@ static char *format_cgroup_memory_limit_comparison(char *buf, size_t l, Unit *u,
 
 void cgroup_context_dump(Unit *u, FILE* f, const char *prefix) {
         _cleanup_free_ char *disable_controllers_str = NULL, *cpuset_cpus = NULL, *cpuset_mems = NULL;
+        _cleanup_free_ char *startup_cpuset_cpus = NULL, *startup_cpuset_mems = NULL;
```
Comment (poettering) [inline src/core/cgroup.c]: merge with previous line?

File: src/core/cgroup.c [hunk 2]
```diff
@@ -1302,8 +1335,10 @@ static void cgroup_context_apply(
         }
 
         if ((apply_mask & CGROUP_MASK_CPUSET) && !is_local_root) {
-                cgroup_apply_unified_cpuset(u, &c->cpuset_cpus, "cpuset.cpus");
-                cgroup_apply_unified_cpuset(u, &c->cpuset_mems, "cpuset.mems");
+                CPUSet *cpus = cgroup_context_allowed_cpus(c, state);
+                CPUSet *mems = cgroup_context_allowed_mems(c, state);
+                cgroup_apply_unified_cpuset(u, cpus, "cpuset.cpus");
```
Comment (poettering) [inline src/core/cgroup.c]: i'd just shorten this to:

```c
cgroup_apply_unified_cpuset(u, cgroup_context_allowed_cpus(c, state), "cpuset.cpus");
```

no need to do this in two steps, the expression doesn't become that complex by that yet

File: src/core/load-fragment.c [hunk 3]
```diff
@@ -3630,7 +3630,10 @@ int config_parse_allowed_cpus(
 
         CGroupContext *c = data;
 
-        (void) parse_cpu_set_extend(rvalue, &c->cpuset_cpus, true, unit, filename, line, lvalue);
+        if (streq(lvalue, "AllowedCPUs"))
+                (void) parse_cpu_set_extend(rvalue, &c->cpuset_cpus, true, unit, filename, line, lvalue);
+        else if (streq(lvalue, "StartupAllowedCPUs"))
+                (void) parse_cpu_set_extend(rvalue, &c->startup_cpuset_cpus, true, unit, filename, line, lvalue);
 
         return 0;
```
Comment (poettering) [inline src/core/load-fragment.c]: i think we should merge config_parse_allowed_cpus() and config_parse_allowed_mems() into one, and then change the `data` parameter to get a pointer to `cgroup_context.cpuset_cpus` instead of `cgroup_context`, and siilar for the other fields.

That would shorten the code quite a bit, and make these lvalue string comparisons unnecessary.

File: man/systemd.resource-control.xml [hunk 4]
```diff
@@ -237,29 +237,42 @@
...
           <para>Restrict processes to be executed on specific memory NUMA nodes. Takes a list of memory NUMA nodes indices
           or ranges separated by either whitespace or commas. Memory NUMA nodes ranges are specified by the lower and upper
           NUMA nodes indices separated by a dash.</para>
 
-          <para>Setting <varname>AllowedMemoryNodes=</varname> doesn't guarantee that all of the memory NUMA nodes will
-          be used by the processes as it may be limited by parent units. The effective configuration is reported as
-          <varname>EffectiveMemoryNodes=</varname>.</para>
+         <para>Setting <varname>AllowedMemoryNodes=</varname> or <varname>StartupAllowedMemoryNodes=</varname> doesn't
+         guarantee that all of the memory NUMA nodes will be used by the processes as it may be limited by parent units.
+         The effective configuration is reported as <varname>EffectiveMemoryNodes=</varname>.</para>
```
Comment (poettering) [inline man/systemd.resource-control.xml]: something is off with the indentation here?

**Review by poettering — CHANGES_REQUESTED** [review]
looks good, some minor changes

File: src/core/cgroup.c (hunk 1 above)
Comment (pdmorrow) [reply src/core/cgroup.c]: Yep, fixed.

File: src/core/cgroup.c (hunk 2 above)
Comment (pdmorrow) [reply src/core/cgroup.c]: Yep, done.

File: man/systemd.resource-control.xml (hunk 4 above)
Comment (pdmorrow) [reply man/systemd.resource-control.xml]: Indeed, sorry about that. Fixed.

File: src/core/load-fragment.c (hunk 3 above)
Comment (pdmorrow) [reply src/core/load-fragment.c]: Yep, sounds much better. Done.

**Comment by poettering** [comment]
lgtm.

(btw, we don't do S-o-b in systemd, that's a kernel thing)

**Comment by pdmorrow** [comment]
> lgtm.
> 
> (btw, we don't do S-o-b in systemd, that's a kernel thing)

Thanks! What is s-o-b ?

**Comment by poettering** [comment]
> Thanks! What is s-o-b ?

Signed-off-by:

**Comment by pdmorrow** [comment]
> > Thanks! What is s-o-b ?
> 
> Signed-off-by:

Ah, gotcha. I'll fixup my git settings for systemd :)

**Comment by yuwata** [comment]
LGTM too.

(@mrc0mmand Ubuntu CIs vanish??)

**Comment by mrc0mmand** [comment]
> LGTM too.
> 
> (@mrc0mmand Ubuntu CIs vanish??)

It seems so, but @bluca is already on it afaik.
